### PR TITLE
Implement client-side querying

### PR DESCRIPTION
**Pull request by brandones**
Builds on https://github.com/PIH/pih-esm-referrals-queue/pull/4 , please merge that one first.

### CODE REVIEW

**Comment by brandones** [comment]
I've been saving the polish for later, but I think it might be the next thing...

Screenshot from 2020-03-31 18-27-03

File: src/referrals-queue/referrals-queue.resource.ts [hunk 1]
```diff
@@ -1,8 +1,13 @@
...
-export function getReferrals(): Observable<Object[]> {
+export function getReferrals({
+  fromDate,
+  toDate,
+  patient = null
+}): Observable<Object[]> {
   return openmrsObservableFetch(
-    `/ws/rest/v1/reportingrest/reportdata/cd7dfde7-764a-4da6-81c2-d5887ed1df51?startDate=2020-01-01&endDate=2020-04-01&locale=en`
+    `/ws/rest/v1/reportingrest/reportdata/cd7dfde7-764a-4da6-81c2-d5887ed1df51?startDate=${fromDate}&endDate=${toDate}&locale=en` +
+      (patient ? "&patient=" + patient : "")
```
Comment (mogoodrich) [inline src/referrals-queue/referrals-queue.resource.ts]: Did you mean to add this in (the patient parameter)?  Looks like it in preparation for server-side filtering and not used?  I guess there's no harm in leaving it around.

File: src/referrals-queue/referrals-queue.component.tsx [hunk 2]
```diff
@@ -1,35 +1,132 @@
...
+
+function prepareQuery(query: string): RegExp[] {
+  return query.split(/\s/).map(token => {
+    const tokenAscii = removeDiacritics(token);
+    return new RegExp("\\b" + tokenAscii, "i");
+  });
+}
+
+function removeDiacritics(str: string) {
+  return str.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
```
Comment (mogoodrich) [inline src/referrals-queue/referrals-queue.component.tsx:130]: I'll assume this does the right thing, but where does this come from?

File: src/referrals-queue/referrals-queue.component.tsx [hunk 3]
```diff
@@ -1,35 +1,132 @@
...
+    prepareQuery(query).every(regexp =>
+      regexp.test(removeDiacritics(referral.patient_name))
+    ) ||
+    new RegExp(query, "i").test(referral.zl_emr_id)
+  );
+}
+
+function prepareQuery(query: string): RegExp[] {
+  return query.split(/\s/).map(token => {
+    const tokenAscii = removeDiacritics(token);
```
Comment (mogoodrich) [inline src/referrals-queue/referrals-queue.component.tsx]: Naming convention question: does str.normalize convert to Ascii?

**Review by mogoodrich — APPROVED** [review]
LGTM... added a few comments, but just clarification questions...

File: src/referrals-queue/referrals-queue.resource.ts (hunk 1 above)
Comment (brandones) [reply src/referrals-queue/referrals-queue.resource.ts]: Oh yeah, that is what happened. I'll clean it up. Thanks!

File: src/referrals-queue/referrals-queue.component.tsx (hunk 2 above)
Comment (brandones) [reply src/referrals-queue/referrals-queue.component.tsx:130]: Oh yeah, I forgot to attribute this! https://stackoverflow.com/a/37511463/1464495

File: src/referrals-queue/referrals-queue.component.tsx (hunk 3 above)
Comment (brandones) [reply src/referrals-queue/referrals-queue.component.tsx]: Oh, good point, it doesn't actually. `removeDiacritics` produces unicode, but with no diacritics. Will rename.

**Pull request merged**